test/framework/machine_helpers.go

```diff
@@ -18,30 +18,104 @@ package framework
 
 import (
 	"context"
+	"fmt"
 
 	. "github.com/onsi/ginkgo"
 	. "github.com/onsi/gomega"
-	"github.com/pkg/errors"
 
+	"github.com/pkg/errors"
 	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
 )
 
-// WaitForMachinesToBeUpgradedInput is the input for WaitForMachinesToBeUpgraded.
-type WaitForMachinesToBeUpgradedInput struct {
+type GetMachinesByMachineDeploymentsInput struct {
+	Lister      Lister
+	ClusterName string
+	Namespace   string
+	Deployment  clusterv1.MachineDeployment
+}
+
+// GetMachinesByMachineDeployments returns Machine objects for a cluster belonging to a machine deployment.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetMachinesByMachineDeployments(ctx context.Context, input GetMachinesByMachineDeploymentsInput) []clusterv1.Machine {
+	opts := byClusterOptions(input.ClusterName, input.Namespace)
+	opts = append(opts, machineDeploymentOptions(input.Deployment)...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, opts...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
+type GetControlPlaneMachinesByClusterInput struct {
+	Lister      Lister
+	ClusterName string
+	Namespace   string
+}
+
+// GetControlPlaneMachinesByCluster returns the Machine objects for a cluster.
+// Important! this method relies on labels that are created by the CAPI controllers during the first reconciliation, so
+// it is necessary to ensure this is already happened before calling it.
+func GetControlPlaneMachinesByCluster(ctx context.Context, input GetControlPlaneMachinesByClusterInput) []clusterv1.Machine {
+	options := append(byClusterOptions(input.ClusterName, input.Namespace), controlPlaneMachineOptions()...)
+
+	machineList := &clusterv1.MachineList{}
+	Expect(input.Lister.List(ctx, machineList, options...)).To(Succeed(), "Failed to list MachineList object for Cluster %s/%s", input.Namespace, input.ClusterName)
+
+	return machineList.Items
+}
+
+// WaitForControlPlaneMachinesToBeUpgradedInput is the input for WaitForControlPlaneMachinesToBeUpgraded.
+type WaitForControlPlaneMachinesToBeUpgradedInput struct {
 	Lister                   Lister
 	Cluster                  *clusterv1.Cluster
 	KubernetesUpgradeVersion string
 	MachineCount             int
 }
 
-// WaitForMachinesToBeUpgraded waits until all machines are upgraded to the correct kubernetes version.
-func WaitForMachinesToBeUpgraded(ctx context.Context, input WaitForMachinesToBeUpgradedInput, intervals ...interface{}) {
```

Comment:
This is technically a breaking change. 
I'm ok for this now, but in future PRs we should start deprecating existing methods and create new one vs changing

Reply:
This method is added in 0.3.4 release, so will keep this change but agree that we should deprecate things first.